### PR TITLE
User GPG Keys client

### LINKED ISSUE #1252: GPG Key Administration API Preview
### Summary

Recently announced on this [blog post](https://developer.github.com/changes/2016-04-04-git-signing-api-preview/) the User API has some enhancements that are now in preview mode.

### Notes on Octokit Implementation

This should be similar to our [`IUserKeysClient`](https://github.com/octokit/octokit.net/blob/master/Octokit/Clients/IUserKeysClient.cs)

#### GPG Keys API

 - [GetAllForCurrent](https://developer.github.com/v3/users/gpg_keys/#list-your-gpg-keys)
 - [Get(int id)](https://developer.github.com/v3/users/gpg_keys/#get-a-single-gpg-key)
 - [Create(NewGpgKey newGpgKey)](https://developer.github.com/v3/users/gpg_keys/#create-a-gpg-key)
 - [Delete(int id)](https://developer.github.com/v3/users/gpg_keys/#delete-a-gpg-key)

Reminder that like all functionality in "preview mode" a custom accepts header needs to be specified (see `AcceptHeaders.cs`)
> To access the API during the preview period, you must provide a custom media type in the Accept header:
> application/vnd.github.cryptographer-preview

#### Naming Things

So now we have two different sorts of keys to manage - Git SSH Keys (what we previously referred to as "Keys") and GPG Keys. This client should be named GpgKeys (or whatever naming style we're using for acronyms) and I'll open a separate task to rename `IUserClient.Keys`.

#### Testing Things

Testing with generated GPG keys may be tricky, but I'm open to ideas around this.

### Working on this issue

As this issue follows issue has been flagged as `up-for-grabs` as it is a reasonably easy issue for a new contributor to pick up, and be guided through the process of contributing to octokit
(that said, if an existing contributor wants to work on this, by all means go for it!)

Please comment here if you want to take this issue.

Please read and follow the [CONTRIBUTING guide](https://github.com/octokit/octokit.net/blob/master/CONTRIBUTING.md) guide and if you need any help, just ask!  Don't be afraid to open a `[WIP]` (work in progress) Pull Request once you've started implementing the changes.

### PR DESCRIPTION
**Pull request by alfhenrik**
- [x] Add `IUserGpgKeysClient.cs` and `UserGpgKeysClient.cs`
 - [x] Add `ApiUrls.GpgKeys`
 - [x] Add `NewGpgKey` request model
 - [x] Add `GpgKey` response model
- [x] Add `IObservableUserGpgKeysClient.cs` and `ObservableUserGpgKeysClient.cs`
- [x] Add unit tests
 - [x] `UserGpgKeysClientTests.cs`
 - [x] `ObservableUserGpgKeysClientTests.cs`
- [x] Add integration tests
 - [x] `UserGpgKeysClientTests.cs`
 - [x] `ObservableUserGpgKeysClientTests.cs`

Fixes #1252

### CODE REVIEW

File: Octokit.Reactive/Clients/ObservableUsersClient.cs [hunk 1]
```diff
@@ -78,6 +80,15 @@ public IObservable<User> Update(UserUpdate user)
...
 
         /// <summary>
+        /// A client for GitHub's UserUser GPG Keys API.
+        /// </summary>
+        /// <remarks>
+        /// See the <a href="https://developer.github.com/v3/users/gpg_keys/">User GPG Keys documentation</a> for more information.
+        /// </remarks>
+        public IObservableUserGpgKeysClient GpgKeys { get; private set; }
+
+
```
Comment (ryangribble) [inline Octokit.Reactive/Clients/ObservableUsersClient.cs]: Extra whitespace line?  (Nitpick, I know!!)

File: Octokit.Tests/Clients/UserGpgKeysClientTests.cs [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+        public class TheGetAllForCurrentMethod
+        {
+            [Fact]
+            public void RequestsTheCorrectUrl()
+            {
+                var connection = Substitute.For<IApiConnection>();
+                var client = new UserGpgKeysClient(connection);
+
+                var expectedUri = ApiUrls.GpgKeys().ToString();
```
Comment (ryangribble) [inline Octokit.Tests/Clients/UserGpgKeysClientTests.cs]: On the unit tests side, we usually hardcode the expectedUri... as a minor safety net to guard against typos in the `ApiUrls` helper class, or inadvertent edits etc

File: Octokit.Tests/Clients/UserGpgKeysClientTests.cs [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+                client.GetAllForCurrent();
+
+                connection.Received().GetAll<GpgKey>(
+                    Arg.Is<Uri>(u => u.ToString() == expectedUri),
+                    Arg.Any<Dictionary<string, string>>(),
+                    Arg.Is<string>(s => s == AcceptHeaders.GpgKeysPreview),
```
Comment (ryangribble) [inline Octokit.Tests/Clients/UserGpgKeysClientTests.cs]: Similarly, we hardcode the accept header value on unit tests so we dont have the unit test and implementation code both relying on the `AcceptHeaders` helper class

File: Octokit.Reactive/Clients/ObservableUserGpgKeysClient.cs [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+        /// <summary>
+        /// Gets all GPG keys for the authenticated user.
+        /// </summary>
+        /// <param name="options">Options for changing the API response</param>
+        /// <remarks>
+        /// See the <a href="https://developer.github.com/v3/users/gpg_keys/#list-your-gpg-keys">API documentation</a> for more information.
+        /// </remarks>
+        /// <returns>A <see cref="IReadOnlyList{GpgKey}"/> of <see cref="GpgKey"/>s for the current user.</returns>
+        public IObservable<GpgKey> GetAllForCurrent(ApiOptions options)
```
Comment (ryangribble) [inline Octokit.Reactive/Clients/ObservableUserGpgKeysClient.cs]: Nice pickup catering for the new API pagination options that were only recently introduced to octokit.net :+1:

File: Octokit.Reactive/Clients/IObservableUsersClient.cs [hunk 5]
```diff
@@ -53,6 +53,15 @@ public interface IObservableUsersClient
         IObservableUserKeysClient Keys { get; }
 
         /// <summary>
+        /// A client for GitHub's UserUser GPG Keys API.
+        /// </summary>
+        /// <remarks>
+        /// See the <a href="https://developer.github.com/v3/users/gpg_keys/">User GPG Keys documentation</a> for more information.
+        /// </remarks>
+        [SuppressMessage("Microsoft.Naming", "CA1704:IdentifiersShouldBeSpelledCorrectly", MessageId = "Gpg")]
+        IObservableUserGpgKeysClient GpgKeys { get; }
```
Comment (ryangribble) [inline Octokit.Reactive/Clients/IObservableUsersClient.cs]: While you were away, I believe it was decided to move towards "singular" property names for sub clients (but plaralised class names) - one such mention is in #1207 .  

So for any new clients being added we should probably follow this convention, thus `IObservableUserGpgKeysClient GpgKey { get; }` (and `IUserGpgKeysClient GpgKey { get; }` on `IUsersClient`)

**Comment by ryangribble** [comment]
Looks good!  I've made a few comments

**Comment by alfhenrik** [comment]
Thanks @ryangribble, I've pushed fixes for your comments.

**Comment by ryangribble** [comment]
sorry @alfhenrik , this just needs a rebase/merge due to some other PR's I merged (probably just `AcceptHeaders.cs` changes)

**Comment by alfhenrik** [comment]
No problem, I'll do that soon

**Comment by alfhenrik** [comment]
Rebased and ready for review again

**Comment by ryangribble** [comment]
Cool, I just sent a PR to your branch to bring your `GpgKeyContext` class inline with the other context helpers we just tweaked

also found the integration tests you added were failing sometimes... turned out there was a missing `await`

**Comment by ryangribble** [comment]
LGTM

**Pull request merged**